src/functional.rs

```diff
@@ -154,62 +134,72 @@ impl<K, V> Tree<K, V> {
-                cmp::Ordering::Less => {
-                    let new_left = n.left.delete(k);
-                    Tree::Node(n.clone_with_children(Rc::new(new_left), Rc::clone(&n.right)))
-                }
-                cmp::Ordering::Equal => match (n.left.as_ref(), n.right.as_ref()) {
-                    (Tree::Leaf, Tree::Leaf) => Tree::Leaf,
-                    (Tree::Leaf, Tree::Node(right)) => Tree::Node(right.clone()),
-                    (Tree::Node(left), Tree::Leaf) => Tree::Node(left.clone()),
-
-                    // If we have two children we have to figure out
-                    // which node to promote. We choose here this node's
-                    // predecessor. That is, the largest node in this node's
-                    // left subtree.
-                    (Tree::Node(left_child), _) => {
-                        let (pred_key, pred_val, new_left) = left_child.delete_largest();
-                        let height = new_left.height().max(n.right.height()) + 1;
-                        Tree::Node(
-                            Node {
-                                height,
-                                key: pred_key,
-                                left: new_left,
-                                right: Rc::clone(&n.right),
-                                value: pred_val,
-                            }
-                            .balance(),
-                        )
-                    }
-                },
-                cmp::Ordering::Greater => {
-                    let new_right = n.right.delete(k);
-                    Tree::Node(n.clone_with_children(Rc::clone(&n.left), Rc::new(new_right)))
-                }
-            },
+        match &self.root {
+            None => Self::new(),
+            Some(n) => Self { root: n.delete(k) },
         }
     }
+}
+
+struct Child<K, V>(Option<Rc<Node<K, V>>>);
+impl<K, V> Clone for Child<K, V> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+impl<K, V> Child<K, V> {
+    fn new() -> Self {
+        Self(None)
+    }
 
-    /// Gets the height of this tree.
     fn height(&self) -> usize {
-        match self {
-            Tree::Leaf => 0,
-            Tree::Node(n) => n.height,
+        match &self.0 {
+            None => 0,
+            Some(n) => n.height,
         }
     }
+
+    fn insert(&self, key: K, value: V) -> Self
+    where
+        K: cmp::Ord,
+    {
+        let new_node = match &self.0 {
+            Some(n) => n.insert(key, value),
+            None => Node::new(key, value),
+        };
+        Self(Some(Rc::new(new_node)))
+    }
+
+    fn find(&self, k: &K) -> Option<&V>
+    where
+        K: cmp::Ord,
+    {
+        match &self.0 {
+            Some(n) => n.find(k),
+            None => None,
+        }
+    }
+
+    fn delete(&self, k: &K) -> Self
+    where
+        K: cmp::Ord,
+    {
+        let new_node = match &self.0 {
+            Some(n) => n.delete(k).map(Rc::new),
```

Comment:
There's an unnecessary allocation here. If `self.0` is a node with no children, and its key doesn't match `k` (so no deletion will occur), we should perform an `Rc::clone` here instead of an `Rc::new`. It's not trivial to do this without some extra information. An example can be seen in https://github.com/mlodato517/rust_bst/commit/60dacba4118f56c01a3c3f0c651e791f89aa83b9.